### PR TITLE
http: added support for %REQUESTED_SERVER_NAME% to extract SNI as a custom header (Fixes #19024)

### LINKED ISSUE #19024: headers op cannot evaluate REQUESTED_SERVER_NAME
reported by Jack Muratore  https://envoyproxy.slack.com/archives/C78M4KW76/p1637074615179900

"%REQUESTED_SERVER_NAME%" can not be evaluated in the headers op, which can be used by tcp_proxy's tunnel and http router. 
I suspect access log cannot use it as well.

The root cause is that this legit var name  REQUESTED_SERVER_NAME has the prefix "REQ" and the parser tried to parse it as a format of 
"REQ(...)".

### PR DESCRIPTION
**Pull request by agrawroh**
This PR adds a new command operator in the header formatter called `%REQUESTED_SERVER_NAME%` which can be used to extract the SNI information from the stream and set it as a custom header. 

**Commit Message:** Added support for `%REQUESTED_SERVER_NAME%` to extract the SNI as a custom header.
**Additional Description:** Adds a new command operator called `%REQUESTED_SERVER_NAME%` in the http header formatter which can be used to extract the SNI information from the stream and set it as a custom header.
**Risk Level:** Low
**Testing:** Unit Tests
**Docs Changes:** Added description on `%REQUESTED_SERVER_NAME%` in the docs.
**Release Notes:** Added
**Platform Specific Features:** N/A

**Signed-off-by:** Rohit Agrawal <rohit.agrawal@databricks.com>

### CODE REVIEW

**Review by KBaichoo — COMMENTED** [review]
Thanks for the contribution! lgtm otherwise

**Review by KBaichoo — DISMISSED** [review]
LGTM thanks! 

@rojkov can you assign or review this as maintainer on call? Thanks

File: test/common/router/header_formatter_test.cc [hunk 1]
```diff
@@ -141,6 +141,17 @@ TEST_F(StreamInfoHeaderFormatterTest, TestFormatWithProtocolVariable) {
   testFormatting(stream_info, "PROTOCOL", "HTTP/1.1");
 }
 
+TEST_F(StreamInfoHeaderFormatterTest, TestFormatWithRequestedServerNameVariable) {
+  NiceMock<Envoy::StreamInfo::MockStreamInfo> stream_info;
+  // Validate for empty Request Server Name
+  testFormatting(stream_info, "REQUESTED_SERVER_NAME", "");
+
+  // Validate for a valid Request Server Name
+  const std::string requested_server_name_ = "foo.bar";
```
Comment (lizan) [inline test/common/router/header_formatter_test.cc:150]: nit: no trailing `_`

File: test/common/router/header_formatter_test.cc [hunk 2]
```diff
@@ -932,6 +944,8 @@ TEST(HeaderParserTest, TestParseInternal) {
   };
 
   NiceMock<Envoy::StreamInfo::MockStreamInfo> stream_info;
+  const std::string requested_server_name_ = "foo.bar";
```
Comment (lizan) [inline test/common/router/header_formatter_test.cc:947]: ditto

**Review by lizan — COMMENTED** [review]
LGTM, just style nits.

File: test/common/router/header_formatter_test.cc (hunk 1 above)
Comment (agrawroh) [reply test/common/router/header_formatter_test.cc]: Fixed. Thanks!

File: test/common/router/header_formatter_test.cc (hunk 2 above)
Comment (agrawroh) [reply test/common/router/header_formatter_test.cc]: Fixed. Thanks!

**Review by lizan — APPROVED** [review]

**Pull request merged**